lib/jsdom/living/helpers/create-event-accessor.js

```diff
@@ -0,0 +1,135 @@
+
+const idlUtils = require("../generated/utils");
+
+const ErrorEvent = require("../generated/ErrorEvent");
+
+exports.appendHandler = function appendHandler(el, eventName) {
+  el.addEventListener(eventName, event => {
+    // https://html.spec.whatwg.org/#the-event-handler-processing-algorithm
+    event = idlUtils.implForWrapper(event);
+
+    const callback = el["on" + eventName];
+    if (callback === null) {
+      return;
+    }
+
+    const specialError = ErrorEvent.isImpl(event) && event.type === "error" &&
+      event.currentTarget.constructor.name === "Window";
+
+    let returnValue = null;
+    if (specialError) {
+      returnValue = callback.call(
+        event.currentTarget, event.message,
+        event.filename, event.lineno, event.colno, event.error
+      );
+    } else {
+      returnValue = callback.call(event.currentTarget, event);
+    }
+
+    if (event.type === "beforeunload") { // TODO: we don't implement BeforeUnloadEvent so we can't brand-check here
+      // Perform conversion which in the spec is done by the event handler return type being DOMString?
+      returnValue = returnValue === undefined || returnValue === null ? null : conversions.DOMString(returnValue);
+
+      if (returnValue !== null) {
+        event._canceledFlag = true;
+        if (event.returnValue === "") {
+          event.returnValue = returnValue;
+        }
+      }
+    } else if (specialError) {
+      if (returnValue === true) {
+        event._canceledFlag = true;
+      }
+    } else if (returnValue === false) {
+      event._canceledFlag = true;
+    }
+  });
+};
+
+// https://html.spec.whatwg.org/#event-handler-idl-attributes
+exports.createEventAccessor = function createEventAccessor(obj, event) {
+  Object.defineProperty(obj, "on" + event, {
+    get() { // https://html.spec.whatwg.org/#getting-the-current-value-of-the-event-handler
+      const value = this._eventHandlers[event];
+      if (!value) {
+        return null;
+      }
+
+      if (value.body !== undefined) {
+        let element;
+        let document;
+        if (this.constructor.name === "Window") {
+          element = null;
+          document = idlUtils.implForWrapper(this.document);
+        } else {
+          element = this;
+          document = element.ownerDocument;
+        }
+        const body = value.body;
+        // TODO: location
+
+        const formOwner = element !== null && element.form ? element.form : null;
+
+        try {
+          // eslint-disable-next-line no-new-func
+          Function(body); // properly error out on syntax errors
+          // Note: this won't execute body; that would require `Function(body)()`.
+        } catch (e) {
+          // TODO: Report the error
+          return null;
+        }
+
+        let fn;
+        const createFunction = vm.isContext(document._global) ? document.defaultView._globalProxy.Function : Function;
+        if (event === "error" && element === null) {
```

Comment:
@Sebmaster, this isn't quite right; error events also get the `with() {}` wrapping, so putting that all in the `else` doesn't work, I think.